action.yml

```diff
@@ -0,0 +1,83 @@
+  headscale-user:
+    description: "Headscale User"
+    required: false
+    default: "github-actions"
+  headscale-preauthkey-expiration:
+    description: "Headscale Preauthkey Expiration"
+    required: false
+    default: "30m"
+  headscale-version:
+    description: "Headscale Version"
+    required: false
+    default: "0.25.1"
+  headscale-os-arch:
+    description: "Headscale OS and Architecture (see https://github.com/juanfont/headscale/releases for options)"
+    required: false
+    default: "linux_amd64"
+
+runs:
+  using: "composite"
+  steps:
+    - id: install-headscale-cli
+      shell: bash
+      env:
+        HEADSCALE_VERSION: ${{ inputs.headscale-version }}
+        HEADSCALE_OS_ARCH: ${{ inputs.headscale-os-arch }}
+      run: |
+        echo "Installing and configuring headscale cli"
+        mkdir -p $HOME/.local/bin
+        curl -L -s -o $HOME/.local/bin/headscale https://github.com/juanfont/headscale/releases/download/v${HEADSCALE_VERSION}/headscale_${HEADSCALE_VERSION}_${HEADSCALE_OS_ARCH}
+        chmod +x $HOME/.local/bin/headscale
+        echo "$HOME/.local/bin" >> "$GITHUB_PATH"
+        # Create an empty config file (workaround for issue https://github.com/juanfont/headscale/issues/2193)
+        mkdir -p $HOME/.headscale && touch $HOME/.headscale/config.yaml
+    
+    - id: install-tailscale-cli
+      shell: bash
+      run: |
+        echo "Installing Tailscale cli"
+        curl -fsSL https://tailscale.com/install.sh | sh
+
+    - id: configure-headscale-user
+      shell: bash
+      env:
+        HEADSCALE_CLI_ADDRESS: ${{ inputs.headscale-cli-address }}
+        HEADSCALE_CLI_API_KEY: ${{ inputs.headscale-cli-api-key }}
+        HEADSCALE_USER: ${{ inputs.headscale-user }}
+        HEADSCALE_PREAUTHKEY_EXPIRATION: ${{ inputs.headscale-preauthkey-expiration }}
+      run: |
+        echo "Creating user $HEADSCALE_USER..."
+        headscale -c $HOME/.headscale/config.yaml user create $HEADSCALE_USER || true
+        echo "Creating preauthkey for user $HEADSCALE_USER..."
+        TAILSCALE_CLI_AUTH_KEY=$(headscale -c $HOME/.headscale/config.yaml preauthkey create --ephemeral --user $HEADSCALE_USER --expiration $HEADSCALE_PREAUTHKEY_EXPIRATION -o json | jq -r .key)
+        # Mask the auth key in the logs
+        echo "::add-mask::$TAILSCALE_CLI_AUTH_KEY"
+        echo "TAILSCALE_CLI_AUTH_KEY=$TAILSCALE_CLI_AUTH_KEY" >> $GITHUB_ENV
+
+    - id: connect-tailscale
+      uses: gacts/run-and-post-run@v1.4.2
+      with:
+        run: |
+          echo "Connecting to Tailscale..."
+          sudo tailscale up \
+            --accept-dns \
+            --accept-routes \
+            --authkey=${TAILSCALE_CLI_AUTH_KEY} \
+            --login-server=https://${{ inputs.headscale-cli-address }} \
+            --operator=${USER} \
+            --timeout=90s 
+          echo "Tailscale status:"
+          tailscale status
+        post: |
+          tailscale down || true
+          echo "Tailscale is down!"
```

Comment:
I used the `gacts/run-and-post-run` action to create a post action that will run `tailscale down` at the end of the job.